### DISCUSSION TITLE
Lesson 7: How Does Solidity Handle Public Struct Getters and Contract Initialization?

### DISCUSSION THREAD

**Discussion by eiyen** [Cyfrin/foundry-full-course-cu#566] [Q&A] [answered]
Hello everyone,

I'm currently working on Lesson 7 of the course and I've come across a concept that is proving difficult to fully grasp. My question revolves around the use of public struct getters in contract initialization.

In my `HelperConfig` contract, I have a public struct called `priceFeedConfig` of type `NetworkConfig`:

```solidity
contract HelperConfig is Script {
    NetworkConfig public priceFeedConfig;

    struct NetworkConfig {
        address priceFeedAddress;
    }
    // Other code...
}
```

As this struct is public, Solidity automatically generates a getter function for it, also named `priceFeedConfig()`.

In my `FundMe` contract, the constructor is expecting an address as an argument:

```solidity
contract FundMe {
    AggregatorV3Interface s_priceFeed;
    
    constructor(address priceFeed) {
        s_priceFeed = AggregatorV3Interface(priceFeed);
    }
    // Other code...
}
```

Now, I'm trying to initialize a `FundMe` instance in my `FundMeScript` contract using the `priceFeedConfig()` getter from an instance of `HelperConfig`:

```solidity
contract FundMeScript is Script{
    function run() external returns(FundMe) {
        HelperConfig helperConfig = new HelperConfig();
        FundMe fundMe = new FundMe(helperConfig.priceFeedConfig());
        // Other code...
    }
}
```

From what I understand, the `priceFeedConfig()` getter should return the `NetworkConfig` struct. However, this does not match with the constructor of `FundMe` which expects an address. Yet, this code compiles and runs without errors. On the other hand, when I attempt to directly access the `priceFeedAddress` field from the getter, like `helperConfig.priceFeedConfig.priceFeedAddress`, I get a compilation error: "Member 'priceFeedAddress' not found or not visible after argument-dependent lookup in function () view external returns (address)."

Can anyone explain why `FundMe(helperConfig.priceFeedConfig())` is valid and how Solidity interprets the `priceFeedConfig()` getter in this scenario? Is the getter returning the `NetworkConfig` struct or the `priceFeedAddress` from that struct?

Thanks in advance for your help!

**Comment by theirrationalone** [comment, marked as answer]
Hey... Patrick discussed this already but once more let me try to explain this for you.... 

```solidity
helperConfig.priceFeedConfig();
```

This chunk of code will return a tuple. And You decided to have only one member which is a **price-feed address** okay... 

Then you created an instance of it In **FundMeScript** using **new** keyword, Okay... 


Solidity is smart enough to format and structure code efficiently, So when you accessed all the members of this: **helperConfig.priceFeedConfig();** as result it returned a tuple in which there has only one member and it was the RValue so it will access values of all members of pricefeedconfig but it has only member which is picefeedAddress so it will return a tuple with one value inside it. 

Here solidity will remove that tuple's parenthesis automatically for simplicity because we have only one value in it. 

And that one value is the address: **priceFeedAddress**, that's exactly what FundMe's constructor needed to have. 

And you did that everything in only one line. You can split them though. 



```solidity
FundMe fundMe = new FundMe(helperConfig.priceFeedConfig());
```
Only for your understanding (pseudo code) :

address priceFeed  = helperConfig.priceFeedConfig();

Now priceFeed = 0x17282priceFeedAddress;

FundMe fundMe = new(because we have a constructor in our contract) FundMe(constructor argument=priceFeed=0x17282priceFeedAddress);

# Access problem of struct members. 

Nothing to do with.... 

That's a limitation in solidity (intentional: storage slots can go in danger). 


Hope it Helps 😅😅

**Reply by theirrationalone** [reply]
And for more info [click here](https://solidity-by-example.org/structs/)

**Reply by eiyen** [reply]
Hello, and thank you for your detailed explanation.

I apologize for not catching this point when Patrick Collins mentioned it. Your explanation, along with the link you provided, helped clarify the issue for me.

From what I understand now, there are two main reasons for this behavior:

1. Since the tuple returned by the `priceFeedConfig()` getter function only contains a single member variable, Solidity automatically unpacks it, assigning the value of that single member variable.

2. Due to considerations of storage slot safety, we cannot directly access the member variables of a struct from outside the contract. Instead, we need to use the getter function generated by Solidity, which returns all member variables of the struct as a tuple. Then, we can destructure this tuple to get the value of the member variable we are interested in.

Again, thank you for your time and help!

**Reply by theirrationalone** [reply]
Thunder ⚡⚡⚡


Wowww you got that right as we do in JS & Python destructuring & Unpacking.
